build_runner/lib/src/generate/build_definition.dart

```diff
@@ -165,17 +128,89 @@ class _Loader {
         _onDelete);
   }
 
-  /// Reads in an [AssetGraph] from disk.
-  Future<AssetGraph> _readAssetGraph(AssetId assetGraphId) async {
-    try {
-      return new AssetGraph.deserialize(
-          JSON.decode(await _options.reader.readAsString(assetGraphId)) as Map);
-    } on AssetGraphVersionException catch (_) {
-      // Start fresh if the cached asset_graph version doesn't match up with
-      // the current version. We don't currently support old graph versions.
-      _logger.info('Throwing away cached asset graph due to version mismatch.');
+  /// Checks that the [_buildActions] are valid based on the
+  /// `_options.writeToCache` setting.
+  void _checkBuildActions() {
+    if (!_options.writeToCache) {
+      final root = _options.packageGraph.root.name;
+      for (final action in _buildActions) {
+        if (action.package != _options.packageGraph.root.name) {
+          throw new InvalidBuildActionException.nonRootPackage(action, root);
+        }
+      }
+    }
+  }
+
+  /// If `_options.writeToCache` is `true` then this returns the all the sources
+  /// found in the cache directory, otherwise it returns an empty set.
+  Future<Set<AssetId>> _getCacheDirSources() async {
+    if (_options.writeToCache) {
+      return await _listGeneratedAssetIds().toSet();
```

Comment:
[nit] no need for `await` here. Also prefer a `get` method to a method prefixed wight 'get'.

Could be

```dart
Future<Set<AssetId>> get _cacheDirSources => _options.writeToCache
    ? _listGeneratedAssetIds().toSet()
    : new Future.value(new Set<AssetId>());
```

or

```dart
Future<Set<AssetId>> get _cacheDirSources async => _options.writeToCache
    ? _listGeneratedAssetIds().toSet()
    : new Set<AssetId>();
```

Reply:
A getter doesn't fit with the style of the rest of the file, I renamed it to `_findCacheDirSources` which is more in line with `_findInputSources` and keeps the style consistent. It also indicates more that real work is being done.